echo_server.py

```diff
@@ -4,29 +4,29 @@
 import threading
 
 
-def handle_client(connection):
-    PRETEND_TO_DO_WORK = 5
-    time.sleep(PRETEND_TO_DO_WORK)
-    data = connection.recv(1024)
-    string = data.decode('utf-8')
-    answer = f'Echo: {string}'
-    connection.send(answer.encode('utf-8'))
-    connection.close()
-
-
-def serve():
-    server = socket.socket()
-    server.bind(('', 3333))
-    server.listen(5)
-
-    while True:
-        connection, peer_address = server.accept()
-        logging.info(f'Connection from {peer_address}')
-        thread = threading.Thread(target=handle_client, args=(connection,))
-        thread.start()
-
+def main():
+    def _handle_client(connection):
+        PRETEND_TO_DO_WORK = 5
+        time.sleep(PRETEND_TO_DO_WORK)
+        data = connection.recv(1024)
+        string = data.decode('utf-8')
+        answer = f'Echo: {string}'
+        connection.send(answer.encode('utf-8'))
+        connection.close()
+
+    def serve():
+        server = socket.socket()
+        server.bind(('', 3333))
+        server.listen(5)
+
```

Comment:
generally speaking, this is well written and it works.

There's a problem when a client calls `.close()` on their socket - this crashes your server.

also, if data is not valid UTF-8, it will crash your server.

these problems are not very hard to solve however.

Good job!